src/Server/EmbeddedClient/EmbeddedClientRunner.cpp

```diff
@@ -0,0 +1,64 @@
+#include <Server/EmbeddedClient/EmbeddedClientRunner.h>
+#include <Common/Exception.h>
+#include "Common/logger_useful.h"
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+void EmbeddedClientRunner::run(const NameToNameMap & envs, const String & starting_query)
+{
+    if (started.test_and_set())
```

Comment:
This potentially sets the `started` flag true but `ThreadFromGlobalPool` (l. 20) may throw. That seems a problem.